crates/ra_hir/src/code_model.rs

```diff
@@ -1362,6 +1362,98 @@ impl Type {
             ty: InEnvironment { value: ty, environment: self.ty.environment.clone() },
         }
     }
+
+    /// Returns a flattened list of all ADTs and Traits mentioned in the type
+    pub fn flattened_type_items(&self, db: &dyn HirDatabase) -> Vec<AdtOrTrait> {
```

Comment:
I'd maybe just re-used `ModuleItem` here instead of introducing super-specific `AdtOrTrait`

Reply:
Did you mean 'ra_hir::code_model::ModuleDef'? If yes, I thought about it, but decided it worth having a special entity for a "type part".

Reply:
>Did you mean 'ra_hir::code_model::ModuleDef'?

Yup

>If yes, I thought about it, but decided it worth having a special entity for a "type part".

Note that in IDE we mostly fold everything into `Definition` kitchen sink anyway. Genrally, in IDE you want to display stuff to the user, and, if all you want is `ToNav`, it doesn't really matter if you try to preserve extra type-safety along the way. 

This is in contrast to compiler layer, where fine-grained enum matching seems like it makes sense.

Reply:
ok. I'll remove AdtOrTrait.

Reply:
Yeah, also I think "ADT or trait" is not a very useful concept -- there's not a lot you can do with these two that you can't do with a general `ModuleDef`.